### PR TITLE
Add ability to add/update abritrary storage configuration for a tenant

### PR DESCRIPTION
**Pull request by MikeLarah**
This adds the `AddConfiguration` to the tenant management service, and corresponding command to the CLI.

### CODE REVIEW

**Comment by jongeorge1** [comment]
I was initially a bit confused by the amount of change that's part of this PR; I'd assumed that "arbitrary configuration" meant that you could literally add anything you like. However it looks like this adds code that's specific to adding storage configuration. Out of interest, was there any particular reason for going down this route? The other option would have been to allow literally anything to be added (since tenant properties are ultimately just key/value pairs).

File: Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/BlobStorageConfigurationItem.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+        /// <inheritdoc/>
+        public override string ContentType => RegisteredContentType;
+
+        /// <inheritdoc/>
+        public override IEnumerable<KeyValuePair<string, object>> AddConfiguration(IEnumerable<KeyValuePair<string, object>> values)
+        {
+            if (this.Configuration.Container == null)
+            {
+                throw new NullReferenceException($"{nameof(this.Configuration.Container)} cannot be null.");
```
Comment (jongeorge1) [inline Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/BlobStorageConfigurationItem.cs:29]: It's valid for `Container` to be null. If not provided, we fall back to that provided in `BlobStorageConfiguration` - see https://github.com/corvus-dotnet/Corvus.Tenancy/blob/2c49f150c9fa093b7ac983175694f16d3436061a/Solutions/Corvus.Azure.Storage.Tenancy/Corvus/Azure/Storage/Tenancy/Internal/TenantCloudBlobContainerFactory.cs#L189

File: Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/CosmosConfigurationItem.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        {
+            if (this.Configuration.DatabaseName == null)
+            {
+                throw new NullReferenceException($"{nameof(this.Configuration.DatabaseName)} cannot be null.");
+            }
+
+            if (this.Configuration.ContainerName == null)
+            {
+                throw new NullReferenceException($"{nameof(this.Configuration.DatabaseName)} cannot be null.");
+            }
```
Comment (jongeorge1) [inline Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/CosmosConfigurationItem.cs:35]: It's valid for either or both of `DatabaseName` and `ContainerName` to be null - we fall back to those provided in the `CosmosContainerDefinition` if they aren't provided. See https://github.com/corvus-dotnet/Corvus.Tenancy/blob/2c49f150c9fa093b7ac983175694f16d3436061a/Solutions/Corvus.Azure.Cosmos.Tenancy/Corvus/Azure/Cosmos/Tenancy/Internal/TenantCosmosContainerFactory.cs#L215

File: Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Exceptions/InvalidConfigurationException.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        {
+            this.Errors = errors.ToArray();
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="InvalidConfigurationException"/> class.
+        /// </summary>
+        /// <param name="info">The <see cref="SerializationInfo"/>.</param>
+        /// <param name="context">The <see cref="StreamingContext"/>.</param>
+        protected InvalidConfigurationException(
```
Comment (jongeorge1) [inline Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Exceptions/InvalidConfigurationException.cs:37]: We normally drop this constructor as well, as we generally accept these exceptions never need to be serialized. @idg10 can provide more info on the background to this decision, as I can't remember the detail right now.

File: Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Internal/TenantManagementService.cs [hunk 4]
```diff
@@ -444,6 +445,57 @@ public async Task<ServiceManifestRequiredConfigurationEntry[]> GetServiceEnrollm
...
 
+        /// <inheritdoc/>
+        public async Task AddConfigurationAsync(ITenant tenant, ConfigurationItem[] configurationItems)
+        {
+            if (tenant == null)
+            {
+                throw new ArgumentNullException(nameof(tenant));
+            }
+
+            tenant.EnsureTenantIsOfType(MarainTenantType.Client, MarainTenantType.Delegated, MarainTenantType.Service);
```
Comment (jongeorge1) [inline Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Internal/TenantManagementService.cs:456]: Why bother? This is basically saying you can attach arbitrary config to any type of tenant.

However, at the moment I can't see why we'd be attaching config to Delegated or Service tenants. I don't think we should ever be directly manipulating delegated tenants, as these are created automatically as part of enrollment on behalf of an existing client tenant. And at the moment, we don't have any scenarios where service tenants are actually used to store configuration. So maybe this should be limited to client tenants only?

File: Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Internal/TenantManagementService.cs [hunk 5]
```diff
@@ -444,6 +445,57 @@ public async Task<ServiceManifestRequiredConfigurationEntry[]> GetServiceEnrollm
...
+            }
+
+            this.logger.LogDebug(
+                "Updating tenant '{tenantName}' with Id '{tenantId}'",
+                tenant.Name,
+                tenant.Id);
+
+            tenant = await this.tenantStore.UpdateTenantAsync(
+                tenant.Id,
+                propertiesToSetOrAdd: propertiesToAddToTenant)
```
Comment (jongeorge1) [inline Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Internal/TenantManagementService.cs:490]: Worth noting that as it stands, this code will overwrite existing config (i.e. it's really "add or update", not "add"). I think we should consider either:
- Renaming the command to reflect this; or,
- Blowing up if you try and update a property that's already set; or,
- Adding an extra parameter to say that you're ok with overwriting existing properties, and blowing up if it's not set and you're trying to set an existing property.

File: Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/TenantManagementServiceExtensions.cs [hunk 6]
```diff
@@ -224,5 +225,36 @@ public static class TenantManagementServiceExtensions
...
+            if (configurationItems == null)
+            {
+                throw new ArgumentNullException(nameof(configurationItems));
+            }
+
+            ITenant tenant = await tenantManagementService.GetTenantOfTypeAsync(
+                tenantId,
+                MarainTenantType.Client,
+                MarainTenantType.Delegated,
+                MarainTenantType.Service).ConfigureAwait(false);
```
Comment (jongeorge1) [inline Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/TenantManagementServiceExtensions.cs:255]: As above, I don't think it's a great idea to allow adding arbitrary config to `MarainTenantType.Delegated` tenants, and not sure why you'd be adding to `MarainTenantType.Service` tenants.

File: Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/BlobStorageConfigurationItem.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+            }
+
+            return values.AddBlobStorageConfiguration(new BlobStorageContainerDefinition(this.Configuration.Container), this.Configuration);
```
[the remark this answers is not in the document]
Comment (MikeLarah) [reply Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/BlobStorageConfigurationItem.cs]: Updated to pass in definition

File: Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/CosmosConfigurationItem.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+            return values.AddCosmosConfiguration(
+                new CosmosContainerDefinition(
+                    this.Configuration.DatabaseName,
+                    this.Configuration.ContainerName,
+                    this.Configuration.PartitionKeyPath),
+                this.Configuration);
```
[the remark this answers is not in the document]
Comment (MikeLarah) [reply Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/CosmosConfigurationItem.cs:32]: Updated to pass in definition

File: Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/TableStorageConfigurationItem.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+        /// <inheritdoc/>
+        public override IEnumerable<KeyValuePair<string, object>> AddConfiguration(IEnumerable<KeyValuePair<string, object>> values)
+        {
+            if (this.Configuration.TableName == null)
+            {
+                throw new NullReferenceException($"{nameof(this.Configuration.TableName)} cannot be null.");
+            }
+
+            return values.AddTableStorageConfiguration(new TableStorageTableDefinition(this.Configuration.TableName), this.Configuration);
```
[the remark this answers is not in the document]
Comment (MikeLarah) [reply Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/TableStorageConfigurationItem.cs]: Updated to pass in definition

File: Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/TableStorageConfigurationItem.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+        /// <inheritdoc/>
+        public override string ContentType => RegisteredContentType;
+
+        /// <inheritdoc/>
+        public override IEnumerable<KeyValuePair<string, object>> AddConfiguration(IEnumerable<KeyValuePair<string, object>> values)
+        {
+            if (this.Configuration.TableName == null)
+            {
+                throw new NullReferenceException($"{nameof(this.Configuration.TableName)} cannot be null.");
```
[the remark this answers is not in the document]
Comment (MikeLarah) [reply Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/TableStorageConfigurationItem.cs]: Removed validation

File: Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/CosmosConfigurationItem.cs (hunk 2 above)
Comment (MikeLarah) [reply Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/CosmosConfigurationItem.cs]: Removed validation

File: Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Internal/TenantManagementService.cs (hunk 4 above)
Comment (MikeLarah) [reply Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Internal/TenantManagementService.cs]: Limited to client tenant

File: Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/TenantManagementServiceExtensions.cs (hunk 6 above)
Comment (MikeLarah) [reply Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/TenantManagementServiceExtensions.cs]: Limited to client tenant

**Comment by MikeLarah** [comment]
@jongeorge1 this is ready for review again. Renamed methods from `AddConfiguration` to `AddOrUpdateStorageConfiguration` to make the intent clearer.

File: Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/ConfigurationItem.cs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+        /// Validates the configuration item.
+        /// </summary>
+        /// <returns>A list of validation errors. An empty list means that the item is valid.</returns>
+        public abstract IList<string> Validate();
+
+        /// <summary>
+        /// Adds the configuration to a property set.
+        /// </summary>
+        /// <param name="values">The property set.</param>
+        /// <returns>The updated property set.</returns>
```
Comment (idg10) [inline Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/ConfigurationItem.cs:34]: I'd like a `<remarks>` section here explaining a bit more about what this is for. I think, from looking through the PR, that this gets called by tenant management during initialization. So presumably it has expectations about what overrides of this method are supposed to do. I can't tell what those are.

File: Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/ConfigurationItemExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    public static class ConfigurationItemExtensions
+    {
+        /// <summary>
+        /// Validates the list of <see cref="ConfigurationItem"/> and throws an
+        /// <see cref="InvalidConfigurationException"/> if any errors are encountered.
+        /// </summary>
+        /// <param name="configurationItems">The configuration items to validate.</param>
+        /// <exception cref="InvalidConfigurationException">The configuration items contain one or more errors.</exception>
+        public static void ValidateAndThrow(
+            this ConfigurationItem[] configurationItems)
```
Comment (idg10) [inline Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/ConfigurationItemExtensions.cs:24]: Is there a reason for this to be an array, and not `IEnumerable<ConfigurationItem>`?

File: Solutions/Marain.TenantManagement.Specs/Marain.TenantManagement.Specs.csproj [hunk 13]
```diff
@@ -27,6 +27,10 @@
     <None Remove="Data\ServiceManifests\WorkflowServiceManifest.jsonc" />
   </ItemGroup>
 
+  <ItemGroup>
+    <SpecFlowObsoleteCodeBehindFiles Remove="Features\TenantManagement\Enrollment - Copy.feature.cs" />
```
Comment (idg10) [inline Solutions/Marain.TenantManagement.Specs/Marain.TenantManagement.Specs.csproj:31]: Not sure this should be here?

File: Solutions/Marain.TenantManagement.Specs/Marain.TenantManagement.Specs.csproj [hunk 14]
```diff
@@ -63,10 +67,23 @@
...
+
   <ItemGroup>
     <None Update="specflow.json">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
   </ItemGroup>
 
+  <ItemGroup>
+    <SpecFlowFeatureFiles Update="Features\TenantManagement\AddOrUpdateStorageConfiguration.feature">
+      <Visible>$(UsingMicrosoftNETSdk)</Visible>
```
Comment (idg10) [inline Solutions/Marain.TenantManagement.Specs/Marain.TenantManagement.Specs.csproj:84]: Is this expected?

Comment (MikeLarah) [reply Solutions/Marain.TenantManagement.Specs/Marain.TenantManagement.Specs.csproj]: No - removed

File: Solutions/Marain.TenantManagement.Specs/Marain.TenantManagement.Specs.csproj (hunk 13 above)
Comment (MikeLarah) [reply Solutions/Marain.TenantManagement.Specs/Marain.TenantManagement.Specs.csproj]: Remove

File: Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/ConfigurationItemExtensions.cs (hunk 12 above)
Comment (MikeLarah) [reply Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/ConfigurationItemExtensions.cs]: Updated

File: Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/ConfigurationItem.cs (hunk 11 above)
Comment (MikeLarah) [reply Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Configuration/ConfigurationItem.cs:40]: Added a remark

**Review by HowardvanRooijen — APPROVED** [review]

**Pull request merged**